### DISCUSSION TITLE
KEEP Alive AIR Desktop (Prevent Screen Server)

### DISCUSSION THREAD

**Discussion by radubirsan** [airsdk/Adobe-Runtime-Support#2728] [General]
I am trying to find a native extension that can keep alive my Herman AIR application on windows during voice calls.
Prevent windows screen server.

**Comment by ajwfrost** [comment]
@marchbold  do you have this?

I wondered whether this API might work:
https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/desktop/NativeApplication.html#systemIdleMode
but it says it's only supported on mobile operating systems!

In case we did want to expand that support to Windows, a quick google found this:
https://stackoverflow.com/questions/463813/programmatically-prevent-windows-screensaver-from-starting
which could be implemented I guess. Quicker to be done via ANE perhaps?

thanks

**Reply by 2jfw** [reply]
I would think that adding these kind of functions natively to AIR would somehow improve the acceptance and hopefully usage of the SDK. Though I do really appreciate the possibility of being able to create and use ANEs, I feel that they are mostly there to overcome missing native AIR SDK features, which _was_ based on Adobe's "unwillingness" to take AIR to the next level (or lets say add overdue features and functionalities).
Harman _can_ -and actually **IS** (!)- doing better ... !

**Reply by marchbold** [reply]
I disagree, I honestly believe Adobe saw ANEs as a massive platform advantage and it is something we now see in all other cross platform frameworks. Adobe definitely dropped the ball on AIR but ANEs weren't born out of "unwillingness" but one of the last great ideas Adobe had, born out of alchemy/crossbridge and the other projects they were running at the peak of AIR :)

**Reply by 2jfw** [reply]
Fair point! Coming and dealing with AIR for desktop only, I find additions like this would be great to have natively being included in AIR; for mobile platforms I do definitely agree that ANEs are especially a suitable and reasonable way to extend beyond limits. Though I see a certain risk of ANEs getting abandoned and no longer maintained then stop working due to platform updates etc (like it happened with the Milkman ANEs) - kind of afraid of that..

**Reply by ajwfrost** [reply]
I think there's two (or three?) arguments in favour of ANEs:
1) for "non-core" functionality which may be specific to different platforms. So e.g. where the behaviour or capabilities of one or two platforms is significantly different from the others, and having this as a generic/core API would be challenging.
2) they enable the functionality of AIR to be enhanced by the community. This is actually something I think is valuable - yes there are issues around maintenance/dependencies perhaps, but when it works, it means you're not dependent on us to add features that may be desired
3) and linked to that, it often means that capabilities/updates can be done quicker by the community rather than waiting for us to get around to it!

On the second point, the other thing we are looking at is enhancing the ANE interface/capabilities so that these can be made to do other things, the key one being to render elements into the display list.

Anyway, this is a bit of a side topic; for this particular feature, I'm not sure why it's mobile-only! It does look possible on macOS too, potentially: https://developer.apple.com/documentation/iokit/kiopmassertiontypepreventuseridledisplaysleep

**Reply by marchbold** [reply]
> On the second point, the other thing we are looking at is enhancing the ANE interface/capabilities so that these can be made to do other things, the key one being to render elements into the display list.

Yes please to rendering into the display list! :)

**Reply by 2jfw** [reply]
> the key one being to render elements into the display list.

Absolutely! I was thinking about that feature a short while back and was wondering how you would deal with `stage.frameRate` and if this would affect the content being displayed? (It's off topic, sry, would it make sense to have a separate thread for this and/or such features to raise questions etc? like "Road Map Talk" or "Planned Features Discussions" or is there already a thread I'm missing ?) 

> they enable the functionality of AIR to be enhanced by the community.

Okay, that is of course the biggest benefit! 👍

**Comment by marchbold** [comment]
@radubirsan I'll have a look around our kiosk projects, I feel like we have created one for this purpose a few years ago. What versions of Windows are you looking to support?

**Reply by radubirsan** [reply]
@marchbold need it for Windows 10 & 11

**Comment by RayGyoe** [comment]
https://github.com/RayGyoe/ANEWinCore
function【keepScreenOn(value:Boolean)】

For reference: Suitable for x86 architecture, of course, x64 support should only require recompilation once

**Reply by radubirsan** [reply]
The ANE dose not work on my Windows 10:

trace("Keep aliv", ANEWinCore.getInstance().isSupported)  // returns true
ANEWinCore.getInstance().keepScreenOn(true)

Screen still turns off after 1 minute inactivity

**Reply by RayGyoe** [reply]
It functions properly on my end... Upon revisiting the document, it indicates the necessity of periodic invocations.


https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-setthreadexecutionstate

`Calling SetThreadExecutionState without ES_CONTINUOUS simply resets the idle timer; to keep the display or system in the working state, the thread must call SetThreadExecutionState periodically.
`

**Reply by radubirsan** [reply]
I can confirm that it works with periodic invocations.

**Comment by radubirsan** [comment]
I think I found a workaround , playback of a mp4 video on loop seams to prevent the screen from turning off.

**Comment by radubirsan** [comment]
@RayGyoe  is there a way off accessing location from [windows ](https://learn.microsoft.com/en-us/windows/uwp/maps-and-location/get-location) using your ANE?

**Reply by RayGyoe** [reply]
after a quick search, the implementation in ane should be relatively simple

[https://learn.microsoft.com/en-us/windows/win32/api/locationapi/nf-locationapi-ilocation-getreportstatus](url)

**Reply by radubirsan** [reply]
Will you implement it in your ANE?
